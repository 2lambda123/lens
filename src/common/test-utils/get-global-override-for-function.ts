/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { Injectable } from "@ogre-tools/injectable";
import { getGlobalOverride } from "./get-global-override";
import { camelCase } from "lodash/fp";

export const getGlobalOverrideForFunction = (
  injectable: Injectable<Function, any, any>,
) =>
  getGlobalOverride(injectable, (di) => (...args: any[]) => {
    console.warn(
      `Tried to invoke a function "${injectable.id}" without override.`,
      "\n\n",
      "The args were:",
      args,
      "\n\n",
      `The context was: "${di.context
        .map((x) => x.injectable.id)
        .join(" -> ")}"`,
    );

    throw new Error(
      `Tried to invoke a function "${
        injectable.id
      }" without override. Add eg. "di.override(${camelCase(
        injectable.id,
      )}Mock)" to the unit test interested in this.`,
    );
  });
