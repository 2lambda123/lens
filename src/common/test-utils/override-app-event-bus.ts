/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { DiContainer, Injectable } from "@ogre-tools/injectable";
import type { AppEvent } from "../app-event-bus/event-bus";
import type { EventEmitter } from "../event-emitter";

export function overrideAppEventBusInjectable(di: DiContainer, appEventBusInjectable: Injectable<EventEmitter<[AppEvent]>, unknown, void>) {
  const emitEvent: (event: AppEvent) => void = jest.fn();

  di.override(appEventBusInjectable, () => ({
    emit: emitEvent,
  }));

  return emitEvent;
}
