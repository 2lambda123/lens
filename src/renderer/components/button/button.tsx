/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import "./button.scss";
import type { ButtonHTMLAttributes } from "react";
import React from "react";
import { cssNames } from "../../utils";
import type { TooltipDecoratorProps } from "../tooltip";
import { withTooltip } from "../tooltip";
import { OnClickDecorated } from "../on-click-decorated/on-click-decorated";

export interface ButtonProps extends ButtonHTMLAttributes<any> {
  label?: React.ReactNode;
  waiting?: boolean;
  primary?: boolean;
  accent?: boolean;
  light?: boolean;
  plain?: boolean;
  outlined?: boolean;
  hidden?: boolean;
  active?: boolean;
  big?: boolean;
  round?: boolean;
  href?: string; // render as hyperlink
  target?: "_blank"; // in case of using @href
}

const prepareProps = (props: ButtonProps) => {
  const {
    waiting, label, primary, accent, plain, hidden, active, big,
    round, outlined, light, children, ...btnProps
  } = props;

  btnProps.className = cssNames("Button", btnProps.className, {
    waiting, primary, accent, plain, active, big, round, outlined, light,
  });

  return btnProps;
};

export const Button = withTooltip((props: ButtonProps) => {
  const { label, children, hidden } = props;

  if (hidden) return null;

  const btnProps = prepareProps(props);

  // render as link
  if (props.href) {
    return (
      <a
        {...btnProps}>
        {label}
        {children}
      </a>
    );
  }

  // render as button
  return (
    <button
      type="button"
      {...btnProps}>
      {label}
      {children}
    </button>
  );
});

export const OpenLensButton = (props: ButtonProps & TooltipDecoratorProps) => {
  const { label, children, hidden } = props;

  if (hidden) return null;

  const buttonProps = prepareProps(props);

  return (
    <OnClickDecorated {...buttonProps} tagName={props.href ? "a" : "button"}>
      {label}
      {children}
    </OnClickDecorated>
  );
};
