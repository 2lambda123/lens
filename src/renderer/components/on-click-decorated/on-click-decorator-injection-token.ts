/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { getInjectionToken } from "@ogre-tools/injectable";
import type { MouseEventHandler } from "react";
import type React from "react";

type OnClick = (toBeDecorated: MouseEventHandler<any>) => (event: React.MouseEvent) => void;

export interface OnClickDecorator {
  onClick: OnClick;
}

export const onClickDecoratorInjectionToken = getInjectionToken<OnClickDecorator>({
  id: "onclick-decorator-injection-token",
});
