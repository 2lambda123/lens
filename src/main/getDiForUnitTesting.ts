/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import glob from "glob";
import { memoize, kebabCase } from "lodash/fp";
import { createContainer } from "@ogre-tools/injectable";

import { setLegacyGlobalDiForExtensionApi } from "../extensions/as-legacy-globals-for-extension-api/legacy-global-di-for-extension-api";
import getElectronAppPathInjectable from "./app-paths/get-electron-app-path/get-electron-app-path.injectable";
import setElectronAppPathInjectable from "./app-paths/set-electron-app-path/set-electron-app-path.injectable";
import appNameInjectable from "./app-paths/app-name/app-name.injectable";
import registerEventSinkInjectable from "../common/communication/register-event-sink.injectable";
import registerChannelInjectable from "./communication/register-channel.injectable";
import { overrideFsFunctions } from "../test-utils/override-fs-functions";

interface DiForTestingOptions {
  doGeneralOverrides?: boolean;
  doIpcOverrides?: boolean;
}

export async function getDiForUnitTesting({ doGeneralOverrides = false, doIpcOverrides = true }: DiForTestingOptions = {}) {
  const di = createContainer();

  setLegacyGlobalDiForExtensionApi(di);

  for (const filePath of getInjectableFilePaths()) {
    const { default: injectableInstance } = await import(filePath);

    di.register({
      id: filePath,
      ...injectableInstance,
      aliases: [injectableInstance, ...(injectableInstance.aliases || [])],
    });
  }

  di.preventSideEffects();

  if (doGeneralOverrides) {
    di.override(getElectronAppPathInjectable, () => (name: string) => `some-electron-app-path-for-${kebabCase(name)}`);

    di.override(setElectronAppPathInjectable, () => () => undefined);
    di.override(appNameInjectable, () => "some-electron-app-name");

    overrideFsFunctions(di);
  }

  if (doIpcOverrides) {
    di.override(registerEventSinkInjectable, () => () => () => undefined);
    di.override(registerChannelInjectable, () => () => () => undefined);
  }

  return di;
}

const getInjectableFilePaths = memoize(() => [
  ...glob.sync("./**/*.injectable.{ts,tsx}", { cwd: __dirname }),
  ...glob.sync("../extensions/**/*.injectable.{ts,tsx}", { cwd: __dirname }),
  ...glob.sync("../common/**/*.injectable.{ts,tsx}", { cwd: __dirname }),
]);
