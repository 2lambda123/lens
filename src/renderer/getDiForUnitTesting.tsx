/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import glob from "glob";
import { memoize } from "lodash/fp";
import { createContainer } from "@ogre-tools/injectable";
import { setLegacyGlobalDiForExtensionApi } from "../extensions/as-legacy-globals-for-extension-api/legacy-global-di-for-extension-api";
import registerEventSinkInjectable from "../common/communication/register-event-sink.injectable";
import registerChannelInjectable from "./communication/register-channel.injectable";
import { overrideFsFunctions } from "../test-utils/override-fs-functions";

interface DiForTestingOptions {
  doGeneralOverrides?: boolean;
  doIpcOverrides?: boolean;
}

export async function getDiForUnitTesting({ doGeneralOverrides = false, doIpcOverrides = true }: DiForTestingOptions = {}) {
  const di = createContainer();

  setLegacyGlobalDiForExtensionApi(di);

  for (const filePath of getInjectableFilePaths()) {
    const { default: injectableInstance } = await import(filePath);

    try {
      di.register({
        id: filePath,
        ...injectableInstance,
        aliases: [injectableInstance, ...(injectableInstance.aliases || [])],
      });
    } catch (error) {
      throw new Error(`Failed to register ${filePath}: ${error}`);
    }
  }

  di.preventSideEffects();

  if (doGeneralOverrides) {
    overrideFsFunctions(di);
  }

  if (doIpcOverrides) {
    di.override(registerEventSinkInjectable, () => () => () => undefined);
    di.override(registerChannelInjectable, () => () => () => undefined);
  }

  return di;
}

const getInjectableFilePaths = memoize(() => [
  ...glob.sync("./**/*.injectable.{ts,tsx}", { cwd: __dirname }),
  ...glob.sync("../common/**/*.injectable.{ts,tsx}", { cwd: __dirname }),
  ...glob.sync("../extensions/**/*.injectable.{ts,tsx}", { cwd: __dirname }),
]);
