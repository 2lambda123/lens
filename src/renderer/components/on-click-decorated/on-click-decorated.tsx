/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { withInjectables } from "@ogre-tools/injectable-react";
import { flow } from "lodash/fp";
import type { HTMLAttributes, MouseEventHandler } from "react";
import React from "react";
import type { OnClickDecorator } from "./on-click-decorator-injection-token";
import {  onClickDecoratorInjectionToken } from "./on-click-decorator-injection-token";

interface Dependencies {
  decorators: OnClickDecorator[];
}

interface ClickDecoratedProps extends HTMLAttributes<any> {
  onClick?: MouseEventHandler<any>;
  tagName: "button" | "a";
}

const NonInjectedOnClickDecorated = ({ decorators, tagName: TagName, onClick, ...props }: Dependencies & ClickDecoratedProps) => {
  const onClickDecorators = decorators.map(decorator => decorator.onClick);

  const decoratedOnClick = onClick ? flow(...onClickDecorators)(onClick) : undefined;

  return <TagName {...props} onClick={decoratedOnClick} />;
};

export const OnClickDecorated = withInjectables<Dependencies, ClickDecoratedProps>(
  NonInjectedOnClickDecorated,

  {
    getProps: (di, props) => ({
      decorators: di.injectMany(onClickDecoratorInjectionToken),
      ...props,
    }),
  },
);
