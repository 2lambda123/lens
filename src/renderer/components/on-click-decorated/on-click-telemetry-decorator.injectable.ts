/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { getInjectable } from "@ogre-tools/injectable";
import type { AppEvent } from "../../../common/app-event-bus/event-bus";
import type { EventEmitter } from "../../../common/event-emitter";
import capitalize from "lodash/capitalize";
import { onClickDecoratorInjectionToken } from "./on-click-decorator-injection-token";
import appEventBusInjectable from "../../../common/app-event-bus/app-event-bus.injectable";

function getEventName(el: HTMLElement, pathname: string, parentLevels = 3) {
  let headers: string[] = [];
  let parent = el;

  const path = pathname.split("/");

  headers.push(capitalize(path[path.length-1]));

  for (let i = 0; i < parentLevels; i++) {
    if (parent.parentElement) {
      parent = parent.parentElement;
    }
  }

  const nodelist = parent.querySelectorAll("h1, h2, h3, h4, h5, h6, .header");

  nodelist.forEach(node => node.textContent && headers.push(node.textContent));

  headers = [...new Set(headers)];

  if (el?.textContent) {
    headers.push(el.textContent);
  }
  const eventName = headers.join(" ");

  return eventName;
}

function captureMouseEvent(eventBus: EventEmitter<[AppEvent]>, event: React.MouseEvent) {
  const name = getEventName(event.target as HTMLElement, window.location.pathname);

  eventBus.emit({
    name,
    action: capitalize(event.type),
    destination: "AutoCapture",
  });
}


const onClickTelemetryDecoratorInjectable = getInjectable({
  id: "on-click-telemetry-decorator",

  instantiate: (di) => ({
    onClick: (toBeDecorated) => {
      return (event) => {
        captureMouseEvent(di.inject(appEventBusInjectable), event);

        return toBeDecorated(event);
      };
    },
  }),

  injectionToken: onClickDecoratorInjectionToken,
});

export default onClickTelemetryDecoratorInjectable;
