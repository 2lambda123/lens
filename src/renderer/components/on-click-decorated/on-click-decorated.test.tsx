/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { getDiForUnitTesting } from "../../getDiForUnitTesting";
import type { DiRender } from "../test-utils/renderFor";
import { renderFor } from "../test-utils/renderFor";
import { OpenLensButton } from "../button";
import React from "react";
import { fireEvent } from "@testing-library/dom";
import type { RenderResult } from "@testing-library/react";
import { onClickDecoratorInjectionToken } from "./on-click-decorator-injection-token";
import type { DiContainer } from "@ogre-tools/injectable";
import { getInjectable } from "@ogre-tools/injectable";
import { OnClickDecorated } from "./on-click-decorated";

describe("<OnClickDecorated />", () => {
  let di: DiContainer;
  let render: DiRender;

  beforeEach(() => {
    di = getDiForUnitTesting({ doGeneralOverrides: true });
    render = renderFor(di);
  });

  it("given A -element", () => {
    const rendered = render(
      <OnClickDecorated
        tagName="a"
        onClick={() => {}}
        data-testid="some-anchor"
      />,
    );

    expect(rendered.baseElement).toMatchSnapshot();
  });


  describe("given button", () => {
    let onClickMock: jest.Mock;
    let firstHigherOrderFunctionMock: jest.Mock;
    let secondHigherOrderFunctionMock: jest.Mock;
    let rendered: RenderResult;

    beforeEach(() => {
      onClickMock = jest.fn();
      firstHigherOrderFunctionMock = jest.fn((x) => x);
      secondHigherOrderFunctionMock = jest.fn((x) => x);

      const someInjectable = getInjectable({
        id: "some-injectable",

        instantiate: () => ({
          onClick: firstHigherOrderFunctionMock,
        }),

        injectionToken: onClickDecoratorInjectionToken,
      });

      const someOtherInjectable = getInjectable({
        id: "some-other-injectable",

        instantiate: () => ({
          onClick: secondHigherOrderFunctionMock,
        }),

        injectionToken: onClickDecoratorInjectionToken,
      });

      di.register(someInjectable, someOtherInjectable);

      rendered = render(
        <OpenLensButton
          primary
          label="Add Custom Helm Repo"
          onClick={onClickMock}
          data-testid="add-custom-helm-repo-button"
        />,
      );
    });

    describe("when button is clicked", () => {
      beforeEach(() => {
        fireEvent.click(rendered.getByTestId("add-custom-helm-repo-button"));
      });

      it("calls the original onClick", () => {
        expect(onClickMock).toHaveBeenCalled();
      });

      it("calls the original onClick with an argument", () => {
        expect(onClickMock).toHaveBeenCalledWith(expect.any(Object));
      });

      it("calls the first higher order function", () => {
        expect(firstHigherOrderFunctionMock).toHaveBeenCalled();
      });

      it("calls the second higher order function", () => {
        expect(secondHigherOrderFunctionMock).toHaveBeenCalled();
      });
    });
  });
});
