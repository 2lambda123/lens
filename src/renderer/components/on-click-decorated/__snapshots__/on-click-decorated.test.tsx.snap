// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OnClickDecorated /> given A -element 1`] = `
<body>
  <div>
    <a
      data-testid="some-anchor"
    />
  </div>
</body>
`;
